AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +632,819 @@ class AzureIoTHub {
-                return this;
+            foreach (optName, optVal in options) {
+                _options[optName] <- optVal;
             }
 
-            if( !_isOpen(_sessions.event) ) {
-                _handlers.onEvent("Cannot receive while disconnected.", null);
-            } else {
-                // TODO: test/handle if receiver already open
-                if ( _isOpen(_receivers.event) ) _receivers.event.close();
-                _openEventReceiver();
-            }
-            return this;
+            _fillTopics(_connStrParsed.DeviceId);
         }
 
-        // PRIVATE FUNCTIONS
-        // ------------------------------------------------------------------------------------
-        // ------------------------------------------------------------------------------------
+        // Opens a connection to Azure IoT Hub.
+        //
+        // Returns:                         Nothing.
+        function connect() {
+            if (_state == AZURE_IOT_CLIENT_STATE_DISCONNECTED) {
+                _log("Connecting...");
 
-        // Open Auth/Event Session/Sender/Receivers
-        // ------------------------------------------------------------------------------------
+                local devPath = "/" + _connStrParsed.DeviceId;
+                local username = _connStrParsed.HostName + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourcePath = "/devices" + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+                local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
+                local sas = AzureIoTHub.SharedAccessSignature.create(
+                    resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
 
-        function _openAuthSession() {
-            _sessions.auth <- _connection.opensession(_amqpAuthSessionStatusHandler.bindenv(this));
-        }
+                local options = {
+                    "username" : username,
+                    "password" : sas
+                };
 
-        function _openAuthReceiver() {
-            _receivers.auth <- _sessions.auth.openreceiver(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthReceiverStatusHandler.bindenv(this), _authReceiverCB.bindenv(this));
-        }
+                local url = "ssl://" + _connStrParsed.HostName;
 
-        function _openAuthSender() {
-            _senders.auth <- _sessions.auth.opensender(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthSenderStatusHandler.bindenv(this));
-        }
+                _state =  AZURE_IOT_CLIENT_STATE_DISCONNECTED | AZURE_IOT_CLIENT_STATE_CONNECTING;
 
-        function _openEventSession() {
-            _sessions.event <- _connection.opensession(_amqpEventSessionStatusHandler.bindenv(this));
+                _mqttclient.connect(url, _connStrParsed.DeviceId, options);
+            } else {
+                try {
```

Comment:
these try-catch statements for invoking the callback are duplicated across the code. won't it be easier to just wrap them into a function and call everywhere? (would save the code space)

Reply:
Done